botocore/handlers.py

```diff
@@ -266,16 +268,67 @@ def add_expect_header(model, params, **kwargs):
             params['headers']['Expect'] = '100-continue'
 
 
-def quote_source_header(params, **kwargs):
-    if params['headers'] and 'x-amz-copy-source' in params['headers']:
+def document_copy_source_form(section, event_name, **kwargs):
+    if 'request-example' in event_name:
+        parent = section.get_section('structure-value')
+        param_line = parent.get_section('CopySource')
+        value_portion = param_line.get_section('member-value')
+        value_portion.clear_text()
+        value_portion.write("'string' or {'Bucket': 'string', "
+                            "'Key': 'string', 'VersionId': 'string'}")
+    elif 'request-params' in event_name:
+        param_section = section.get_section('CopySource')
+        type_section = param_section.get_section('param-type')
+        type_section.clear_text()
+        type_section.write(':type CopySource: str or dict')
+        doc_section = param_section.get_section('param-documentation')
+        doc_section.clear_text()
+        doc_section.write(
+            "The name of the source bucket, key name of the source object, "
+            "and optional version ID of the source object.  You can either "
+            "provide this value as a string or a dictionary.  The "
+            "string form is {bucket}/{key} or "
+            "{bucket}/{key}?versionId={versionId} if you want to copy a "
+            "specific version.  You can also provide this value as a "
+            "dictionary.  The dictionary format is recommended over "
+            "the string format because it is more explicit.  The dictionary "
+            "format is: {'Bucket': 'bucket', 'Key': 'key', 'VersionId': 'id'}."
+            "  Note that the VersionId key is optional and may be omitted."
+        )
+
+
+def support_dict_copy_source(params, context, **kwargs):
+    source = params.get('CopySource')
+    if source is None or not isinstance(source, dict):
+        return
+    source_dict = params['CopySource']
+    try:
+        bucket = source_dict['Bucket']
```

Comment:
It doesn't need to be encoded.  I think it was done that way initially because it doesn't matter.  The characters allowed for a bucket name don't change when urlencoded.